common/blobstore/s3store/config_test.go

```diff
@@ -0,0 +1,126 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package s3store
+
+import (
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+	"testing"
+)
+
+type ConfigSuite struct {
+	*require.Assertions
+	suite.Suite
+}
+
+func TestConfigSuite(t *testing.T) {
+	suite.Run(t, new(ConfigSuite))
+}
+
+func (s *ConfigSuite) SetupTest() {
+	s.Assertions = require.New(s.T())
+}
+
+func (s *ConfigSuite) TestValidate() {
+	testCases := []struct {
+		config  *Config
+		isValid bool
+	}{
+		{
+			config:  &Config{},
+			isValid: false,
+		},
+		{
+			config: &Config{
+				DefaultBucket: BucketConfig{},
+			},
+			isValid: false,
+		},
+		{
+			config: &Config{
+				DefaultBucket: BucketConfig{
+					Name: "test-default-bucket-name",
+				},
+			},
+			isValid: false,
+		},
+		{
+			config: &Config{
+				Region: "eu-west-1",
+				DefaultBucket: BucketConfig{
+					Name: "test-default-bucket-name",
+				},
+			},
+			isValid: true,
+		},
+		{
+			config: &Config{
+				DefaultBucket: BucketConfig{
+					Name: "test-default-bucket-name",
+				},
+			},
+			isValid: false,
+		},
+		{
+			config: &Config{
+				Region: "eu-west-1",
+				DefaultBucket: BucketConfig{
+					Name: "test-default-bucket-name",
+				},
+			},
+			isValid: true,
+		},
+		{
+			config: &Config{
+				Region: "eu-west-1",
+				DefaultBucket: BucketConfig{
+					Name: "test-default-bucket-name",
+				},
+				CustomBuckets: []BucketConfig{
+					{},
+				},
+			},
+			isValid: false,
+		},
+		{
+			config: &Config{
+				Region: "eu-west-1",
+				DefaultBucket: BucketConfig{
+					Name: "test-default-bucket-name",
+				},
+				CustomBuckets: []BucketConfig{
+					{
+						Name: "test-custom-bucket-name",
+					},
+				},
+			},
+			isValid: true,
+		},
+	}
+
+	for _, tc := range testCases {
+		if tc.isValid {
```

Comment:
nit: nice to use subtests.. so 
```
t.Run(testname, func (t *testing.T) {
    if tc.IsValue ...
})
```

Reply:
I am okay with the existing pattern. This is how we write other config tests throughout code base.

Reply:
Sure, but we should move to subtests as they are more grokkable on failures - subtests are newer than when the cadence code base started so were not used initially is my guess.. Also, they allow you to parallelize tests to make them faster if we want to make them faster, rather than running them serial..